runtime/jcl/common/java_lang_Class.cpp

```diff
@@ -1821,6 +1821,7 @@ Java_java_lang_Class_getNestHostImpl(JNIEnv *env, jobject recv)
 			 * not specify a nest, then it is considered to belong to its own nest and this is returned as
 			 * the host */
 			nestHost = clazzToUse;
+			clazzToUse->nestHost = nestHost;
```

Comment:
Can this be moved into `loadAndVerifyNestHost`?  It's cleaner if that's the only code that sets `nestHost`.  A special case here makes it easy to overlook and harder to reason about nestmate code in general

Reply:
So the existing code in loadAndVerifyNestHost() is already setting clazz->nestHost whenever J9_VISIBILITY_ALLOWED is returned. I cached nestHost to clazzToUse->nestHost here because I noticed one issue with existing nest mate implementation in LambdaNestedInnerTest.

If a class (could be a normal class, not necessarily a hidden class) has problem finding its nest host, i.e. loadAndVerifyNestHost() failed, this class is in its own nest. The test is doing something like assert(Clazz.getNestHost() ==  Clazz.getNestMembers()). Our existing code before this change has Clazz.getNestHost() == Clazz and Clazz.getNestMembers() == null, which makes the test fail.

Reply:
That seems like a bug in our implementation of `Java_java_lang_Class_getNestMembersImpl` in that it returns null if the host isn't visible

Reply:
> That seems like a bug in our implementation of Java_java_lang_Class_getNestMembersImpl in that it returns null if the host isn't visible

Yes, I agree.

Reply:
Are you removing this and making the change to `Java_java_lang_Class_getNestMembersImpl`?

Reply:
Would adding this case to Class#getNestMembers remove the need for this line?  It keeps the same behaviour in the natives and adds a small change in the java method
```
iff --git a/jcl/src/java.base/share/classes/java/lang/Class.java b/jcl/src/java.base/share/classes/java/lang/Class.java
index b13dfac0d6..0ad27fca06 100644
--- a/jcl/src/java.base/share/classes/java/lang/Class.java
+++ b/jcl/src/java.base/share/classes/java/lang/Class.java
@@ -4758,6 +4758,10 @@ public Class<?>[] getNestMembers() throws LinkageError, SecurityException {
        }
 
        Class<?>[] members = getNestMembersImpl();
+       if (members == null) {
+               /* NestHost wasn't accessible - class is in a nest with itself */
+               return new Class<?>[] { this };
+       }
        /* Skip security check for the Class object that belongs to the nest consisting only of itself */
        if (members.length > 1) {
                SecurityManager securityManager = System.getSecurityManager();
```

Reply:
May need to put the getNestMembersImpl() into a try catch block, I am getting error/exception if we cannot find the nesthost.

Reply:
As long as we match the RI and the specification, I'm OK with this approach

Reply:
We'll need to validate the behaviour on both JDK11 & JDK15 to ensure we're consistent with both releases

Reply:
Looking at the javadoc of Class.getNestMembers(), it says LinkageError (NoClassDefFoundError is a subclass of LinkageError) will be thrown if it has problem loading or validating a nest member or the nest host. So I think the try catch block is not the way to go.

Reply:
OK, tried removing the the code to cache nesthost to clazzToUse->nestHost, making change in the native Java_java_lang_Class_getNestMembersImpl. I still cannot get this test passed, it failed at a later place (on NoClassDefFoundError). So here is the thing:

A hidden class is added to the nest of its host class, the host class cannot find its nest host.
Both hostClass.getNestHost() and hiddenClass().getNestHost() is called in the test, but the test does not expect the second call hiddenClass().getNestHost() to throw NoClassDefFoundError. I guess RI must cached the result of first call, so hiddenClass().getNestHost() directly return hostClass, it is like what I am doing in Java_java_lang_Class_getNestHostImpl() here.

Reply:
What happens if the test calls `hostClass.getNestHost()` a second time?  By caching the result, I think you avoid the exception on the second call which isn't expected.

Reply:
It also changes the behaviour of `loadAndVerifyNestHost` as it won't throw exceptions on subsequent calls.

Reply:
> What happens if the test calls hostClass.getNestHost() a second time? By caching the result, I think you avoid the exception on the second call which isn't expected.

In Class.java, there is a field `nestHost`, so getNestHost() will call into getNestHostImpl only once for the same class. The second call will return the cached nestHost and won't throw exception. 

https://github.com/eclipse/openj9/blob/master/jcl/src/java.base/share/classes/java/lang/Class.java#L4696

> It also changes the behaviour of loadAndVerifyNestHost as it won't throw exceptions on subsequent calls.

Yes, for the same class, it won't throw exception on subsequent calls, like Class.getNestHost(). I haven't find a way to not cache the nesthost and make the test pass.

Reply:
What does the RI do for the second call to `hostClass.getNestHost()` with the invalid nestHost?

Reply:
Interesting, I made 2 calls of hostClass.getNestHost(). I don't even see RI throw an exception on the first call, it directly returns hostclass itself. The second call has no exception either.

Reply:
OK, I've checked the javadoc of Class.getNestHost(), it only says SecurityException can be thrown, not matter it is first or second call.

Reply:
I wrote the following test case and I've tested it on both the RI and OpenJ9 using JDK14 (don't have a 15 build of the RI handy):

```java
public class Out {

public static class In {
public static void main(String[] args) {
	try {
		System.out.println(In.class.getNestHost());
		System.out.println(java.util.Arrays.toString(In.class.getNestMembers()));
	} catch (Throwable t) {
		t.printStackTrace();
	}
	try {
		System.out.println(In.class.getNestHost());
		System.out.println(java.util.Arrays.toString(In.class.getNestMembers()));
	} catch (Throwable t) {
		t.printStackTrace();
	}
}
}
}
```

Compile it with: `javac Out.java`
Delete Out.class: `rm Out.class`
Run with `java Out\$In`

With both VMs I get similar output:
```
# java Out\$In
class Out$In
java.lang.NoClassDefFoundError: Unable to load nest-host class (Out) of Out$In
	at java.base/java.lang.Class.getNestMembers0(Native Method)
	at java.base/java.lang.Class.getNestMembers(Class.java:4146)
	at Out$In.main(Out.java:7)
Caused by: java.lang.NoClassDefFoundError: Out
	... 3 more
class Out$In
java.lang.NoClassDefFoundError: Unable to load nest-host class (Out) of Out$In
	at java.base/java.lang.Class.getNestMembers0(Native Method)
	at java.base/java.lang.Class.getNestMembers(Class.java:4146)
	at Out$In.main(Out.java:13)
Caused by: java.lang.NoClassDefFoundError: Out
	... 3 more
```

So `getNestHost()` never throws for an invalid host but `getNestMembers()` does.  Is there behaviour different for the RI on JDK15?

Reply:
Yes, behaviour changed in JDK15. I run your test and get the same result on RI and openJ9 (with the change in this PR):

```
java Out\$In
class Out$In
[class Out$In]
class Out$In
[class Out$In]
```

Reply:
I guess I need to put this line `clazzToUse->nestHost = nestHost;` inside `#if JAVA_SPEC_VERSION >= 15`, so that all the change in the PR has effect only on JDK15 and up.

Reply:
Updated to put this line inside `#if JAVA_SPEC_VERSION >= 15`

Reply:
The behaviour is at least consistent for reflection if neither `getNestHost` or `getNestMembers` throws.  I would still prefer to see the write to `clazz->nestHost` happen in `loadAndVerifyNestHost` so that it's consistently handled in that code.  That's the single point that should be writing `nestHost` for runtime resolves (ie: apart from classes that start in their own nests)

Reply:
I checked the latest version of javadoc on Class.getNestMembers() in Java 15 from RI, they removed throwing LinkageError there. (Seems our javadoc needs to updated as well, I was checking our javadoc, which still says Class.getNestMembers() throws Linkage error). For java 15 and up, I changed the code to set clazz->nestHost to itself (host class for hidden class)  inside loadAndVerifyNestHost() if there is problem finding the nest host.

Reply:
PR to update Javadoc: https://github.com/eclipse/openj9/pull/10554